RiotSwiftUI/Modules/Authentication/QRLogin/Common/Service/MatrixSDK/QRLoginService.swift

```diff
@@ -233,14 +233,83 @@ class QRLoginService: NSObject, QRLoginServiceProtocol {
+        
+        let cryptoResult = await withCheckedContinuation { continuation in
+            session.enableCrypto(true) { response in
+                continuation.resume(returning: response)
+            }
+        }
+        
+        guard case .success = cryptoResult else {
+            await teardownRendezvous(state: .failed(error: .rendezvousFailed))
+            return
+        }
+        
+        MXLog.debug("[QRLoginService] Session created, sending device details")
+        guard let requestData = try? JSONEncoder().encode(QRLoginRendezvousPayload(type: .loginProgress,
+                                                                                   outcome: .success,
+                                                                                   deviceId: session.myDeviceId,
+                                                                                   deviceKey: session.crypto.deviceEd25519Key)),
               case .success = await rendezvousService.send(data: requestData) else {
             await teardownRendezvous(state: .failed(error: .rendezvousFailed))
             return
         }
         
-        state = .completed(session: session)
+        MXLog.debug("[QRLoginService] Wait for cross-signing details")
+        guard case let .success(data) = await rendezvousService.receive(),
+              let responsePayload = try? JSONDecoder().decode(QRLoginRendezvousPayload.self, from: data),
+              responsePayload.outcome == .verified,
+              let verifiyingDeviceId = responsePayload.verifyingDeviceId,
+              let verifyingDeviceKey = responsePayload.verifyingDeviceKey else {
+            await teardownRendezvous(state: .failed(error: .rendezvousFailed))
+            return
+        }
+        MXLog.debug("[QRLoginService] Received cross-signing details \(responsePayload)")
+        
+        guard let verifyingDeviceInfo = session.crypto.device(withDeviceId: verifiyingDeviceId, ofUser: session.myUserId) else {
+            await teardownRendezvous(state: .failed(error: .rendezvousFailed))
+            return
+        }
+        
+        MXLog.debug("[QRLoginService] Found verifying device info \(verifyingDeviceInfo)")
+        
+        var securityCompleted = false
+        if verifyingDeviceInfo.fingerprint == verifyingDeviceKey {
+            MXLog.debug("[QRLoginService] Locally marking the existing device as verified \(verifyingDeviceInfo)")
+            await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
+                session.crypto.setDeviceVerification(.verified, forDevice: verifiyingDeviceId, ofUser: session.myUserId) {
+                    MXLog.debug("[QRLoginService] Marked the existing device as verified")
+                    
+                    MXLog.debug("[QRLoginService] Recovering secrets Through the existing device")
+                    session.crypto.crossSigning.requestPrivateKeys(toDeviceIds: [verifiyingDeviceId]) {
```

Comment:
This method will recover only cross-signing keys (SSK, USK, MSK) but not the key backup key.
Calling `session.crypto.requestAllPrivateKeys()` will get the key of the key backup AND will restore all keys from the backup.

Reply:
Turns out that's already happening as part of this delayed request [here](https://github.com/matrix-org/matrix-ios-sdk/blob/develop/MatrixSDK/Crypto/MXCrypto.m#L1490) which also means that we don't need to requests secrets at all, marking the device as trusted is enough.